### PR TITLE
add shift years function for data elements and intervals

### PR DESCRIPTION
**Pull request by zlister**
Updates the DateTimes on the patients when the measurement period is adjusted.

Pull requests into cqm-models require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] Internal ticket for this PR: https://jira.mitre.org/browse/BONNIE-1848
- [x] Internal ticket links to this PR
- [x] Code diff has been done and been reviewed
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass
- [x] If applicable, the library version number in `package.json` and `cqm-models.gemspec` has been updated
- [x] Cqm-execution fixtures have been updated with the update_cqm_execution_fixtures.sh script inside server-scripts using this branch in the cqm-converter

**Bonnie Reviewer:**

Name: @losborne 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

**Cypress Reviewer:**

Name:
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

### CODE REVIEW

File: app/models/qdm/basetypes/data_element.rb [hunk 1]
```diff
@@ -89,6 +89,53 @@ def shift_facility_location_dates(facility_location, seconds)
       facility_location['locationPeriod'][:high] = (facility_location['locationPeriod'][:high].to_time + seconds).to_datetime
     end
 
+    # The necessary reason for this function is to avoid a problem when shifting
+    # past a year that is a leap year. February 29th dates are handled by moving
+    # back to the 28th in non leap years
+    def shift_years(year) 
+      # Iterate over fields
```
Comment (losborne) [inline app/models/qdm/basetypes/data_element.rb]: The line below is clear enough that this comment isn't needed

File: app/models/qdm/basetypes/data_element.rb [hunk 2]
```diff
@@ -89,6 +89,53 @@ def shift_facility_location_dates(facility_location, seconds)
...
+      fields.keys.each do |field|
+        # Check if field is a DateTime
```
Comment (losborne) [inline app/models/qdm/basetypes/data_element.rb]: Same as above (comment not needed)

File: app/models/qdm/basetypes/data_element.rb [hunk 3]
```diff
@@ -89,6 +89,53 @@ def shift_facility_location_dates(facility_location, seconds)
...
+        if send(field).is_a? DateTime
+          if send(field).month == 2 && send(field).day == 29 && !Date.leap?(year.to_i)
+            send(field + '=', send(field).change(year: year.to_i, day: 28))
+          else
+            send(field + '=', send(field).change(year: year.to_i))
+          end
+        end
+        # Check if field is an Interval
```
Comment (losborne) [inline app/models/qdm/basetypes/data_element.rb]: Comment not needed (and already stale, since there is also DataElement check below)

File: app/models/qdm/basetypes/data_element.rb [hunk 4]
```diff
@@ -89,6 +89,53 @@ def shift_facility_location_dates(facility_location, seconds)
...
+    # The necessary reason for this function is to avoid a problem when shifting
+    # past a year that is a leap year. February 29th dates are handled by moving
+    # back to the 28th in non leap years
+    def shift_years(year) 
+      # Iterate over fields
+      fields.keys.each do |field|
+        # Check if field is a DateTime
+        if send(field).is_a? DateTime
+          if send(field).month == 2 && send(field).day == 29 && !Date.leap?(year.to_i)
+            send(field + '=', send(field).change(year: year.to_i, day: 28))
```
Comment (losborne) [inline app/models/qdm/basetypes/data_element.rb]: This is setting the year on the data element to the year passed in, which is passed in like this from Bonnie:

```
  def shift_years(measure, year)
    CQM::Patient.by_user_and_hqmf_set_id(current_user, measure.hqmf_set_id).each do |patient|
      patient.qdmPatient.dataElements.each do |data_element|
        data_element.shift_years(year)
      end
      patient.save!
    end
  end
```

Instead, it should shift the years by the difference between the data element's current year and the old measurement period.  We can calculate that in the `measurement_period` function in Bonnie before updating it on the measure, then pass it into this function.

This function currently will only be accurate if the year was the same as the old measurement period.

This function should instead take a parameter that says how many years to shift by (positive or negative), and then adjust the dates by adding or subtracting that value from the data element's current year.

File: app/models/qdm/basetypes/data_element.rb [hunk 5]
```diff
@@ -89,6 +89,53 @@ def shift_facility_location_dates(facility_location, seconds)
...
+          else
+            send(field + '=', send(field).change(year: year.to_i))
+          end
+        end
+        # Check if field is an Interval
+        if (send(field).is_a? Interval) || (send(field).is_a? DataElement)
+          send(field + '=', send(field).shift_years(year))
+        end
+
+        # Special case for facility locations
```
Comment (losborne) [inline app/models/qdm/basetypes/data_element.rb]: The if/elsif below should be handled polymorphically rather than in the base class:

To do this, remove the block below but extend `shift_years` on FacilityLocation to provide the functionality of `shift_facility_location_years` below after calling the dataElement super `shift_years`.  Then extend `shift_years` in Encounter Performed to include a loop over each `facilityLocation` that calls that function.

**Comment by losborne** [comment]
In addition to changes requested above, tests are needed to cover the added functions

File: app/models/qdm/basetypes/data_element.rb (hunk 4 above)
Comment (zlister) [reply app/models/qdm/basetypes/data_element.rb]: Ah yes, that is my bad, I will adjust that now.

File: app/models/qdm/basetypes/interval.rb [hunk 6]
```diff
@@ -32,6 +32,27 @@ def shift_dates(seconds)
...
+               end
+      end
+      if (@high.is_a? DateTime) || (@high.is_a? Time)
+        high_shift = @high.year + year_shift
+        @high = if @high.month == 2 && @high.day == 29 && !Date.leap?(high_shift)
+                  @high.change(year: high_shift, day: 28)
+                else
+                  @high.change(year: high_shift)
+                end
+        @high = @high.year > 9999 ? @high.change(year: 9999) : @high
```
Comment (losborne) [inline app/models/qdm/basetypes/interval.rb]: We should add a similar line to the low block to make the lowest possible year to be 1 (and a unit test to show that it does so)

Comment (zlister) [reply app/models/qdm/basetypes/interval.rb]: Yep that makes sense. I also added this logic to `data_element.rb` as well.

File: app/models/qdm/encounter_performed.rb [hunk 7]
```diff
@@ -17,5 +17,16 @@ class EncounterPerformed < DataElement
     field :qdmCategory, type: String, default: 'encounter'
     field :qdmStatus, type: String, default: 'performed'
     field :qdmVersion, type: String, default: '5.4'
+
+    def shift_years(year_shift)
```
Comment (jbradl11) [inline app/models/qdm/encounter_performed.rb]: This is a generated file. If we want this function to be added to this class, it will need to be added in the generator.

Comment (losborne) [reply app/models/qdm/encounter_performed.rb]: Hmm, when i run `./bin/validate_generator.sh` locally it exits with 1, but somehow travis didn't fail?

File: app/models/qdm/basetypes/data_element.rb [hunk 8]
```diff
@@ -89,6 +86,53 @@ def shift_facility_location_dates(facility_location, seconds)
...
+        if field == 'facilityLocations'
+          facilityLocations = send(field)
+          shiftedFacilityLocations = []
+          facilityLocations.each do |location|
+            # Need to convert to a QDM::FacilityLocation because it is being passed in as a Hash
+            facilityLocation = QDM::FacilityLocation.new(location)
+            shift_facility_location_years(facilityLocation, year_shift)
+            shiftedFacilityLocations << facilityLocation
+          end
+          facilityLocations = shiftedFacilityLocations
```
Comment (jbradl11) [inline app/models/qdm/basetypes/data_element.rb]: Do you need to perform this assignment? I think you can just use `shiftedFacilityLocations`.

Comment (zlister) [reply app/models/qdm/basetypes/data_element.rb]: Oh gosh, yep. Will change that.
